File: modules/context-native/ios/PrivacySafeLogger.swift
import Foundation
#if canImport(OSLog)
import OSLog
#endif

enum PrivacySafeLogError: Error, Equatable {
  case unsafeEvent
  case unsafeField
  case unsafeValue

  var stableCode: String {
    switch self {
    case .unsafeEvent: "UNSAFE_LOG_EVENT"
    case .unsafeField: "UNSAFE_LOG_FIELD"
    case .unsafeValue: "UNSAFE_LOG_VALUE"
    }
  }
}

enum PrivacySafeLogValue: Equatable {
  case string(String)
  case integer(Int64)
  case decimal(Double)
}

struct PrivacySafeLogger {
  typealias Sink = (String) -> Void

  private static let allowedEvents: Set<String> = [
    "inbox_scan",
    "import_completed",
    "import_failed",
    "ocr_completed",
    "pdf_probe_completed",
  ]
  private static let allowedCodes: Set<String> = [
    "INBOX_SCAN_FAILED",
    "NATIVE_ADAPTER_UNAVAILABLE",
    "NATIVE_MANIFEST_INVALID",
    "NATIVE_OCR_RESULT_INVALID",
    "NATIVE_PDF_RESULT_INVALID",
    "OCR_IMAGE_DECODE_FAILED",
    "OCR_RECOGNITION_FAILED",
    "SHARE_COPY_FAILED",
    "SHARE_IMPORT_FAILED",
    "SHARE_IMPORT_EVENT_INVALID",
  ]
  private static let allowedEngines: Set<String> = [
    "apple-vision",
    "ml-kit-latin",
    "ml-kit-chinese",
    "pdfkit",
    "pdf-renderer",
  ]
  private static let allowedKeys: Set<String> = [
    "code",
    "count",
    "bytes",
    "durationMs",
    "version",
    "engine",
    "anonymousId",
  ]
  private static let versionPattern = try! NSRegularExpression(
    pattern: #"^[0-9]+(?:\.[0-9]+){0,3}$"#
  )
  private static let anonymousIdPattern = try! NSRegularExpression(
    pattern: #"^(?:[0-9a-f]{64}|[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12})$"#
  )

  private let sink: Sink

  init(sink: @escaping Sink = PrivacySafeLogger.systemSink) {
    self.sink = sink
  }

  func log(event: String, fields: [String: PrivacySafeLogValue] = [:]) throws {
    sink(try Self.serialize(event: event, fields: fields))
  }

  static func serialize(
    event: String,
    fields: [String: PrivacySafeLogValue] = [:]
  ) throws -> String {
    guard allowedEvents.contains(event) else { throw PrivacySafeLogError.unsafeEvent }
    guard Set(fields.keys).isSubset(of: allowedKeys) else {
      throw PrivacySafeLogError.unsafeField
    }

    var record: [String: Any] = ["event": event]
    for (key, value) in fields {
      switch (key, value) {
      case ("code", .string(let string)) where allowedCodes.contains(string):
        record[key] = string
      case ("engine", .string(let string)) where allowedEngines.contains(string):
        record[key] = string
      case ("version", .string(let string)) where
        string.utf8.count <= 32 && matches(versionPattern, value: string):
        record[key] = string
      case ("anonymousId", .string(let string)) where
        matches(anonymousIdPattern, value: string):
        record[key] = string
      case ("count", .integer(let number)) where number >= 0:
        record[key] = number
      case ("bytes", .integer(let number)) where number >= 0:
        record[key] = number
      case ("durationMs", .integer(let number)) where number >= 0:
        record[key] = number
      case ("durationMs", .decimal(let number)) where number.isFinite && number >= 0:
        record[key] = number
      default:
        throw PrivacySafeLogError.unsafeValue
      }
    }

    guard JSONSerialization.isValidJSONObject(record),
          let serialized = String(
            data: try JSONSerialization.data(withJSONObject: record, options: [.sortedKeys]),
            encoding: .utf8
          ) else {
      throw PrivacySafeLogError.unsafeValue
    }
    return serialized
  }

  private static func matches(_ expression: NSRegularExpression, value: String) -> Bool {
    let range = NSRange(value.startIndex..<value.endIndex, in: value)
    return expression.firstMatch(in: value, range: range)?.range == range
  }

  private static func systemSink(_ serialized: String) {
    #if canImport(OSLog)
    privacySafeOSLogger.info("\(serialized, privacy: .public)")
    #endif
  }
}

#if canImport(OSLog)
private let privacySafeOSLogger = Logger(
  subsystem: "com.aicontextpack",
  category: "privacy-safe"
)
#endif

File: modules/context-native/android/src/main/java/com/aicontextpack/nativebridge/PrivacySafeLogger.kt
package com.aicontextpack.nativebridge

import android.util.Log
import java.math.BigDecimal

internal class PrivacySafeLogException(val stableCode: String) : IllegalArgumentException(stableCode)

internal class PrivacySafeLogger(
  private val sink: (String) -> Unit = { serialized ->
    Log.i("AIContextPack", serialized)
  },
) {
  fun log(event: String, fields: Map<String, Any> = emptyMap()) {
    sink(serialize(event, fields))
  }

  companion object {
    private val allowedEvents = setOf(
      "inbox_scan",
      "import_completed",
      "import_failed",
      "ocr_completed",
      "pdf_probe_completed",
    )
    private val allowedCodes = setOf(
      "INBOX_SCAN_FAILED",
      "NATIVE_ADAPTER_UNAVAILABLE",
      "NATIVE_MANIFEST_INVALID",
      "NATIVE_OCR_RESULT_INVALID",
      "NATIVE_PDF_RESULT_INVALID",
      "OCR_IMAGE_DECODE_FAILED",
      "OCR_RECOGNITION_FAILED",
      "SHARE_COPY_FAILED",
      "SHARE_IMPORT_FAILED",
      "SHARE_IMPORT_EVENT_INVALID",
    )
    private val allowedEngines = setOf(
      "apple-vision",
      "ml-kit-latin",
      "ml-kit-chinese",
      "pdfkit",
      "pdf-renderer",
    )
    private val allowedKeys = setOf(
      "code",
      "count",
      "bytes",
      "durationMs",
      "version",
      "engine",
      "anonymousId",
    )
    private val versionPattern = Regex("^[0-9]+(?:\\.[0-9]+){0,3}$")
    private val anonymousIdPattern = Regex(
      "^(?:[0-9a-f]{64}|[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12})$",
    )

    fun serialize(event: String, fields: Map<String, Any> = emptyMap()): String {
      if (event !in allowedEvents) throw PrivacySafeLogException("UNSAFE_LOG_EVENT")
      if (!allowedKeys.containsAll(fields.keys)) {
        throw PrivacySafeLogException("UNSAFE_LOG_FIELD")
      }

      val record = sortedMapOf("event" to quoted(event))
      fields.forEach { (key, value) ->
        record[key] = when (key) {
          "code" -> approvedString(value, allowedCodes)
          "engine" -> approvedString(value, allowedEngines)
          "version" -> approvedString(value) {
            it.length <= 32 && versionPattern.matches(it)
          }
          "anonymousId" -> approvedString(value, anonymousIdPattern::matches)
          "count", "bytes" -> nonNegativeInteger(value)
          "durationMs" -> nonNegativeNumber(value)
          else -> throw PrivacySafeLogException("UNSAFE_LOG_FIELD")
        }
      }
      return record.entries.joinToString(prefix = "{", postfix = "}") { (key, value) ->
        "${quoted(key)}:$value"
      }
    }

    private fun approvedString(value: Any, allowed: Set<String>): String =
      approvedString(value, allowed::contains)

    private fun approvedString(value: Any, predicate: (String) -> Boolean): String {
      if (value !is String || !predicate(value)) {
        throw PrivacySafeLogException("UNSAFE_LOG_VALUE")
      }
      return quoted(value)
    }

    private fun nonNegativeInteger(value: Any): String {
      val number = when (value) {
        is Byte -> value.toLong()
        is Short -> value.toLong()
        is Int -> value.toLong()
        is Long -> value
        else -> throw PrivacySafeLogException("UNSAFE_LOG_VALUE")
      }
      if (number < 0) throw PrivacySafeLogException("UNSAFE_LOG_VALUE")
      return number.toString()
    }

    private fun nonNegativeNumber(value: Any): String {
      val number = when (value) {
        is Byte -> value.toDouble()
        is Short -> value.toDouble()
        is Int -> value.toDouble()
        is Long -> value.toDouble()
        is Float -> value.toDouble()
        is Double -> value
        else -> throw PrivacySafeLogException("UNSAFE_LOG_VALUE")
      }
      if (!number.isFinite() || number < 0) {
        throw PrivacySafeLogException("UNSAFE_LOG_VALUE")
      }
      return BigDecimal.valueOf(number).stripTrailingZeros().toPlainString()
    }

    private fun quoted(value: String): String = "\"$value\""
  }
}
